lib/9pfs/9pfs_vnops.c

```diff
@@ -447,21 +447,31 @@ static int uk_9pfs_read(struct vnode *vp, struct vfscore_file *fp,
 	if (!uio->uio_resid)
 		return 0;
 
-	iov = uio->uio_iov;
-	while (!iov->iov_len) {
-		uio->uio_iov++;
-		uio->uio_iovcnt--;
-	}
+	while (i < uio->uio_iovcnt) {
+		iov = &uio->uio_iov[i];
+		if (!iov->iov_len) {
+			i++;
+			continue;
+		}
 
-	rc = uk_9p_read(dev, fid, uio->uio_offset,
-			   iov->iov_len, iov->iov_base);
-	if (rc < 0)
-		return -rc;
+		rc = uk_9p_read(dev, fid, uio->uio_offset,
+				iov->iov_len, iov->iov_base);
+		if (unlikely(rc < 0))
+			return -rc;
+		if (!rc)
+			break;
 
-	iov->iov_base = (char *)iov->iov_base + rc;
-	iov->iov_len -= rc;
-	uio->uio_resid -= rc;
-	uio->uio_offset += rc;
+		iov->iov_base = (char *)iov->iov_base + rc;
+		iov->iov_len -= rc;
+		uio->uio_offset += rc;
+		uio->uio_resid -= rc;
+
+		/* Test for file offset overflow */
+		UK_ASSERT(uio->uio_offset <= __OFF_MAX - rc);
+
+		/* Test for negative remaining number of bytes to process */
+		UK_ASSERT(uio->uio_resid >= 0);
```

Comment:
Asserts testing for overflow/underflow conditions should always test before actually performing the operation. Especially, the first `UK_ASSERT` does not make sense after the fact. Comments are not necessary, from my point of view, as these are obvious checks.

```suggestion
		UK_ASSERT(uio->uio_offset <= __OFF_MAX - rc);
		UK_ASSERT(uio->uio_resid >= rc);
		UK_ASSERT(uio->iov_len >= rc);

		iov->iov_base = (char *)iov->iov_base + rc;
		iov->iov_len -= rc;
		uio->uio_offset += rc;
		uio->uio_resid -= rc;
```

Reply:
You're totally right, my bad for this. Especially for the first assert.